cpp/oneapi/dal/backend/primitives/reduction/reduction_dpc.cpp

```diff
@@ -0,0 +1,151 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <type_traits>
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_dpc.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_rw_dpc.hpp"
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw_dpc.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+inline sycl::event reduce_rm_rw(sycl::queue& q,
+                                const ndview<Float, 2, ndorder::c>& input,
+                                ndview<Float, 1>& output,
+                                const BinaryOp& binary,
+                                const UnaryOp& unary,
+                                const event_vector& deps) {
+    ONEDAL_ASSERT(input.has_data());
+    ONEDAL_ASSERT(output.has_mutable_data());
+    ONEDAL_ASSERT(0 <= input.get_dimension(1));
+    ONEDAL_ASSERT(0 <= input.get_dimension(0));
+    ONEDAL_ASSERT(input.get_dimension(0) <= output.get_dimension(0));
+    using kernel_t = reduction_rm_rw<Float, BinaryOp, UnaryOp>;
```

Comment:
If the function named `reduce`, probably its better to name the class as `reducer`?

Reply:
I assume that verb is more suitable here